src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/SplashScreen.cs

```diff
@@ -274,160 +275,154 @@ private void Fadeout_Tick(object unused, EventArgs args)
-                    UnsafeNativeMethods.WIC.CreateStream(pImagingFactory, out pIStream));
+                using ComScope<IWICImagingFactory> scope = new(pImagingFactory);
 
-                UnsafeNativeMethods.HRESULT.Check(
-                    UnsafeNativeMethods.WIC.InitializeStreamFromMemory(pIStream, pImgBuffer, (uint)cImgBufferLen));
+                // Use the WIC stream class to wrap the unmanaged pointer
+                using ComScope<IWICStream> pIStream = new(null);
+                pImagingFactory->CreateStream(pIStream).ThrowOnFailure();
+                pIStream.Value->InitializeFromMemory(buffer).ThrowOnFailure();
 
                 // Create an object that will decode the encoded image
-                Guid vendor = Guid.Empty;
-                UnsafeNativeMethods.HRESULT.Check(
-                    UnsafeNativeMethods.WIC.CreateDecoderFromStream(pImagingFactory, pIStream,
-                                                                    ref vendor, 0, out pDecoder));
+                using ComScope<IWICBitmapDecoder> pDecoder = new(null);
+                pImagingFactory->CreateDecoderFromStream(
+                    (IStream*)(void*)pIStream,
+                    IID.NULL(),
+                    WICDecodeOptions.WICDecodeMetadataCacheOnDemand,
+                    pDecoder).ThrowOnFailure();
 
                 // Get the frame from the decoder. Most image formats have only a single frame, in the case
                 // of animated gifs we are ok with only displaying the first frame of the animation.
-                UnsafeNativeMethods.HRESULT.Check(
-                    UnsafeNativeMethods.WIC.GetFrame(pDecoder, 0, out pDecodedFrame));
+                using ComScope<IWICBitmapFrameDecode> pDecodedFrame = new(null);
+                pDecoder.Value->GetFrame(0, pDecodedFrame).ThrowOnFailure();
 
-                UnsafeNativeMethods.HRESULT.Check(
-                    UnsafeNativeMethods.WIC.CreateFormatConverter(pImagingFactory, out pBitmapSourceFormatConverter));
+                using ComScope<IWICFormatConverter> pBitmapSourceFormatConverter = new(null);
+                pImagingFactory->CreateFormatConverter(pBitmapSourceFormatConverter).ThrowOnFailure();
 
                 // Convert the image from whatever format it is in to 32bpp premultiplied alpha BGRA
-                Guid pixelFormat = UnsafeNativeMethods.WIC.WICPixelFormat32bppPBGRA;
-                UnsafeNativeMethods.HRESULT.Check(
-                    UnsafeNativeMethods.WIC.InitializeFormatConverter(pBitmapSourceFormatConverter, pDecodedFrame,
-                                                                      ref pixelFormat, 0 /*DitherTypeNone*/, IntPtr.Zero,
-                                                                      0, UnsafeNativeMethods.WIC.WICPaletteType.WICPaletteTypeCustom));
+                pBitmapSourceFormatConverter.Value->Initialize(
+                    (IWICBitmapSource*)(void*)pDecodedFrame,
```

Comment:
```suggestion
                    (IWICBitmapSource*)pDecodedFrame,
```

Reply:
Not going to change these ones. This used to not work- I'm going to change `ComScope` to prevent this from working.

For direct interop, these are not actually interfaces, they are structs. As such they cannot have actual inheritance and there is no way to capture the relationship, making these effectively reinterpret casts. I want it to be in your face when you cast away from the `T*`.

Reply:
@JeremyKuhne Yeah that's perfectly fine, I thought it might be the case. Thanks for explaining the intent on that though :)